### PR TITLE
[devbox.json] Don't create nixpkgs commit hash

### PR DESCRIPTION
**Pull request by mikeland86**
## Summary

This stops adding nixpkgs commit hash to devbox.json. User can still add their own. In the future I can imagine us deprecating and eventually removing this field.

e.g. something like this:

```bash
Warning: nixpkgs commit hash in devbox.json is deprecated. Please run `devbox config upgrade` to remove and use lockfile to resolve package versions.
```

Turned `Nixpkgs` into a pointer so that it doesn't show up in devbox.json

## How was it tested?

```
devbox init
devbox add curl php@latest
devbox run which curl
devbox run which php
```

### CODE REVIEW

File: internal/impl/devbox.go [hunk 1]
```diff
@@ -181,7 +173,7 @@ func (d *Devbox) ShellPlan() (*plansdk.ShellPlan, error) {
 		return nil, err
 	}
 
-	nixpkgsInfo := plansdk.GetNixpkgsInfo(d.cfg.Nixpkgs.Commit)
+	nixpkgsInfo := plansdk.GetNixpkgsInfo(d.cfg.NixPkgsCommitHash())
```
Comment (savil) [inline internal/impl/devbox.go:176]: This doesn't look safe if an empty-string is passed into GetNixpkgsInfo:

> mirrorURL := fmt.Sprintf("%s/nixos/nixpkgs/archive/%s.tar.gz", baseURL, commitHash)

File: internal/impl/devbox.go [hunk 2]
```diff
@@ -243,7 +235,7 @@ func (d *Devbox) Shell(ctx context.Context) error {
 		WithShellStartTime(shellStartTime),
 	}
 
-	shell, err := NewDevboxShell(d.cfg.Nixpkgs.Commit, opts...)
+	shell, err := NewDevboxShell(d.cfg.NixPkgsCommitHash(), opts...)
```
Comment (savil) [inline internal/impl/devbox.go:238]: similarly, here, we need to change `nix.FlakeNixpkgs` to work with empty commit hashes

File: internal/impl/devbox.go [hunk 3]
```diff
@@ -146,7 +138,7 @@ func (d *Devbox) ConfigHash() (string, error) {
 }
 
 func (d *Devbox) NixPkgsCommitHash() string {
-	if hash := d.cfg.Nixpkgs.Commit; hash != "" {
+	if hash := d.cfg.NixPkgsCommitHash(); hash != "" {
```
Comment (savil) [inline internal/impl/devbox.go:141]: do we need line 146 below, even with this change?

File: internal/impl/global.go [hunk 4]
```diff
@@ -142,7 +142,7 @@ func (d *Devbox) pullGlobalFromPath(path string) error {
 }
 
 func (d *Devbox) addFromPull(pullCfg *Config) error {
-	if pullCfg.Nixpkgs.Commit != plansdk.DefaultNixpkgsCommit {
+	if pullCfg.NixPkgsCommitHash() != plansdk.DefaultNixpkgsCommit {
```
Comment (savil) [inline internal/impl/global.go:145]: 1. Why do we still need this change?
2. Do we still require `plansdk.DefaultNixpkgsCommit` or could we get rid of it too?

**Review by savil — COMMENTED** [review]
some questions inline:

**Comment by mikeland86** [comment]
@savil I think I addressed all the issues by defaulting `NixPkgsCommitHash()` if there's nothing in the config. I also removed the code comparing hashes in global pull which is no longer needed.

**Review by savil — APPROVED** [review]
Cool, I think the code works.

We could even say we require `plansdk.DefaultNixpkgsCommit` until we update our `examples/` to use versioned packages.

However, assuming we do update the `examples/` to use versioned packages, then longer term what role does `plansdk.DefaultNixpkgsCommit` play? Do we envision needing to keep it around?

**Comment by mikeland86** [comment]
`DefaultNixpkgsCommit` is currently used for mkShell in flake.nix.tmpl (in some cases). We also want to continue to be backwards compatible with unversioned packages so we may need it for that as well. 

I'm unsure we will get rid of it for a while, but it will be only used in limited situations.

**Pull request merged**